Android: Add selection dialog (drop down/combo box)

srifqi opened the pull request:
**Goal of the PR**
This PR adds selection dialog on Android that will be shown when a player taps a drop down (combo box).

![Minetest Android selection dialog](https://github.com/minetest/minetest/assets/4017302/bbc95c9b-08c3-4c67-8f79-0ce45f0d522e)

<details>
<summary>Videos</summary>

https://github.com/minetest/minetest/assets/4017302/45c566dc-cca2-4ea6-b486-fdbbf45fcf0b

https://github.com/minetest/minetest/assets/4017302/4fb5315d-4ba4-4241-aea0-6fd48238a616

</details>

**How does the PR work?**
In `GUIFormSpecMenu::getAndroidUIInput()` and `GUIModalMenu::preprocessEvent()`, combo boxes are also processed.

This PR also refactors and simplifies interaction between the Java (Android) part and the C++ part (see `porting_android.cpp/h`). See the commit message for more technical information.

**Does this relate to a goal in [the roadmap](https://github.com/minetest/minetest/blob/master/doc/direction.md)?**
Yes, this PR tries to improve user experiences.

**If not a bug fix, why is this PR needed? What usecases does it solve?**
The Irrlicht's combo box (drop down) is small and hard to navigate on Android (touch-screen devices). Hard to scroll worsen this issue.

## To do

This PR is a Ready for Review.

## How to test

To test the change in the main menu,
1. Run Minetest on a touch-screen device.
2. Open any dialog/setting that has selections/drop downs/combo boxes.
3. Tap it.
4. A selection dialog should show up.
5. Tap a different selection.
6. The formspec should update with new data.
7. Repeat step 3 to 5, but with the same selection (as given).
8. The formspec should stay the same.
9. Repeat step 3 to 4.
10. Tap outside the dialog.
11. The formspec should stay the same.

To test the change inside a world,
1. Run Minetest on a touch-screen device.
2. Create a world (optional) and use a mod that has selections/drop downs/combo boxes.
3. Enter the world.
4. Open the dialog from the mod.
5. Try the same as testing in the main menu above.

**Android: APK builds**
Included in these builds is [the Mail mod](https://content.minetest.net/packages/mt-mods/mail/). Type `/mail` in the chat to open the formspec. These builds use https://github.com/minetest/irrlicht/pull/241 for their IrrlichtMT library.
- [armeabi-v7a](https://drive.google.com/file/d/1PS2wtbOZI6dxzG5xtps_PIZRrs4Tg2eu/view)
- [arm64-v8a](https://drive.google.com/file/d/1ZN3j9q7LKdhT9OaoMI7vRFx1moLglcea/view)
- [x86](https://drive.google.com/file/d/13rw5QI9ac3bX_alnJda7YJaCLDU3f2JN/view)
- [x86_64](https://drive.google.com/file/d/1QAYgt0XNcwjXofZZgft_F7sn1-aig2sp/view)

<details>
<summary>Old builds (without the PR in the IrrlichtMT)</summary>

- [armeabi-v7a](https://drive.google.com/file/d/1JtBIqEZ_yykSBPVyjcdQGZJDz3ScM0qn/view)
- [arm64-v8a](https://drive.google.com/file/d/1LtE8sALurnDHNwU9tiKnQtOf8xyVP17o/view)
- [x86](https://drive.google.com/file/d/1OdaE85rq4DZkjNagRDVP49Q027YIYC15/view)
- [x86_64](https://drive.google.com/file/d/17Fn511yhdNxaixVASN5oCbn9jpY64d-1/view)
</details>

grorp commented:
I tested your arm64 APK. Most of the time, it worked fine, but a few times, I somehow accidentally broke it. Here's a video:

https://github.com/minetest/minetest/assets/82708541/3200cfe4-27b3-4ac9-8261-4df45f728ed3

First, the Android dialog didn't show and I was shown a standard Minetest dropdown instead. Then, I tried to select something in that dropdown and the Android dialog opened. When I closed the Android dialog, the whole settings dialog closed with it.

srifqi commented:
> (video)
> First, the Android dialog didn't show and I was shown a standard Minetest dropdown instead. Then, I tried to select something in that dropdown and the Android dialog opened. When I closed the Android dialog, the whole settings dialog closed with it.

This is caused by Irrlicht opening the combo box if we clicked/tapped outside the formspec area while the combo box is still focused (also happens on desktop). We can try to remove focus from the combo box by focusing on its parent element after selecting the item to prevent the issue.

```diff
diff --git a/src/gui/guiFormSpecMenu.cpp b/src/gui/guiFormSpecMenu.cpp
index ea89dcdfa..bcda1f655 100644
--- a/src/gui/guiFormSpecMenu.cpp
+++ b/src/gui/guiFormSpecMenu.cpp
@@ -3503,6 +3503,7 @@ bool GUIFormSpecMenu::getAndroidUIInput()
				change.GUIEvent.Element = 0;
				change.GUIEvent.EventType = EGET_COMBO_BOX_CHANGED;
				dropdown->getParent()->OnEvent(change);
+				Environment->setFocus(dropdown->getParent());
			}
		}
```

The dialog close is caused by passing an <kbd>Esc</kbd> keyboard event to the combo box, but since the combo box is already closed, the event is passed to the dialog and closes it. There is no method to close a combo box list (or at least check its status).

oong819 commented:
Minetest formspec on Android are very unTouchscreenFriendly and somewhat small (made for cursor)

Should we consider create a functiom that read formspec and create Android widget accordingly?

oong819 commented:
oh i have no problem with the PR, its work fine. i just got reminded of irrlicht scrollview X_X

srifqi commented:
I rebased and squashed this PR.

The grorp's PR has been merged on IrrlichtMT, too.

grorp commented:
This is still waiting for a new IrrrlichtMT revision.

srifqi commented:
I rebased this PR to include the newest IrrlichtMT version.

grorp reviewed (changes requested):
This PR makes dropdowns much more usable on Android. Thank you!

While testing, I noticed one problem: You can no longer cancel the text input completely by pressing the back button.

**Before this PR**

https://github.com/minetest/minetest/assets/82708541/4f179357-3e42-4a79-b912-5cbd3d124ff0

- If you press the "Done" button without changing the pre-filled text, the pre-filled text is submitted.
- If you press the back button, no submission happens.


**After this PR**

https://github.com/minetest/minetest/assets/82708541/b12ebbb9-b7ea-4131-965f-a0d0ff4aef84

- If you press the "Done" button without changing the pre-filled text, the pre-filled text is submitted.
- If you press the back button, the pre-filled text is submitted. (<- wrong!)

srifqi commented on src/porting_android.h:
```diff
@@ -51,11 +51,19 @@ void initializePathsAndroid();
-void showInputDialog(const std::string &acceptButton,
-					const std::string &hint, const std::string &current, int editType);
+void showTextDialog(const std::string &acceptButton, const std::string &hint,
+		const std::string &current, int editType);
```
Done.

`showTextDialog` → `showTextInputDialog`
`showSelectionDialog` → `showSelectionInputDialog`

srifqi commented:
> While testing, I noticed one problem: You can no longer cancel the text input completely by pressing the back button.

Thank you for noticing it.

I added another `enum` to represent input dialog states:
1. is currently shown,
2. has its input sent, and
3. is cancelled/dismissed.

-------

I also rebased this PR to current `master` and added/fixed some comments. I might squash these commits later.

grorp reviewed (changes requested):
The text input dialog *sometimes* doesn't properly set the value:

https://github.com/minetest/minetest/assets/82708541/12f69cd1-8b54-4f03-bd44-32d0447c5cbc

So far, I've only observed on the first try after starting Minetest.

Shouldn't `getAndroidUIInput` return void instead of bool? Its return value is never used as far as I can see, and it always returns false.

srifqi commented:
I have rebased and squashed this PR.

> The text input dialog _sometimes_ doesn't properly set the value:
> (video)
> 
> So far, I've only observed on the first try after starting Minetest.

Oh, that is true. The event propagation (to C++ side) seems to be late by one/two events. I will try to look into it.

> Shouldn't `getAndroidUIInput` return void instead of bool? Its return value is never used as far as I can see, and it always returns false.

I agree.

srifqi commented on src/porting_android.cpp:
```diff
@@ -183,6 +183,27 @@ void showInputDialog(const std::string &acceptButton, const std::string &hint,
+			"([Ljava/lang/String;I)V");
+
+	FATAL_ERROR_IF(showdialog == nullptr,
+			"porting::showComboBoxDialog unable to find Java showSelectionDialog method");
```
Fixed.

srifqi commented:
> The text input dialog _sometimes_ doesn't properly set the value:

I can not find the actual cause, but I found that delaying `GameActivity::getDialogMessage()` (making it longer to response) somehow fixes the issue (or at least making it rarer for an already not-often case). This does not explain the value being late by one/two events.

```diff
diff --git a/android/app/src/main/java/net/minetest/minetest/GameActivity.java b/android/app/src/main/java/net/minetest/minetest/GameActivity.java
index f2ff09f3d..d5b057ffa 100644
--- a/android/app/src/main/java/net/minetest/minetest/GameActivity.java
+++ b/android/app/src/main/java/net/minetest/minetest/GameActivity.java
@@ -180,6 +180,7 @@ public class GameActivity extends NativeActivity {
 
 	public String getDialogMessage() {
 		inputDialogState = DialogState.DIALOG_CANCELED;
+		for (int i = 0; i < 1e5;) i ++;
 		return messageReturnValue;
 	}
 
```

-------

By the way, that issue also happens in the recent `master` branch. I still has not been able to reproduce the issue in the 5.8.0 version.

grorp commented:
> By the way, that issue also happens in the recent `master` branch. I still has not been able to reproduce the issue in the 5.8.0 version.

Hmmm, if you say it doesn't happen on 5.8.0 ... could you try whether it's fixed if you revert 00d9d96e48ec3337c415dbc600b0b0800eaecfad?

srifqi commented:
> could you try whether it's fixed if you revert 00d9d96e48ec3337c415dbc600b0b0800eaecfad?

The issue seems to be caused by that commit.

We need a better heuristic for `shouldRender()`:

https://github.com/minetest/minetest/blob/2c44620e5e344ca2b4acaaaa5408502332ca5aff/src/client/renderingengine.h#L141-L150

... or just always render the GUI:

```diff
diff --git a/src/gui/guiEngine.cpp b/src/gui/guiEngine.cpp
index 4bb7b041f..43ea4d2b2 100644
--- a/src/gui/guiEngine.cpp
+++ b/src/gui/guiEngine.cpp
@@ -267,3 +267,2 @@ void GUIEngine::run()
 	while (m_rendering_engine->run() && (!m_startgame) && (!m_kill)) {
-		if (RenderingEngine::shouldRender()) {
 		// check if we need to update the "upper left corner"-text
@@ -295,3 +294,2 @@ void GUIEngine::run()
 		driver->endScene();
-		}
 
```

grorp commented:
> The issue seems to be caused by that commit.

Alright, sorry for reporting it on your PR. Feel free to ignore the issue for this PR.

> We need a better heuristic for `shouldRender()`

I have an idea for that: Perhaps checking for `!Paused` instead of `Focused && !Paused` in Irrlicht fixes the issue. I'll try tomorrow.

> ... or just always render the GUI:
> 
> ```diff
> diff --git a/src/gui/guiEngine.cpp b/src/gui/guiEngine.cpp
> index 4bb7b041f..43ea4d2b2 100644
> --- a/src/gui/guiEngine.cpp
> +++ b/src/gui/guiEngine.cpp
> @@ -267,3 +267,2 @@ void GUIEngine::run()
>  	while (m_rendering_engine->run() && (!m_startgame) && (!m_kill)) {
> -		if (RenderingEngine::shouldRender()) {
>  		// check if we need to update the "upper left corner"-text
> @@ -295,3 +294,2 @@ void GUIEngine::run()
>  		driver->endScene();
> -		}
>  
> ```

`guiEngine.cpp` is only for the mainmenu, so this won't affect in-game GUIs. `guiEngine.cpp` not the "GUI engine", but the "engine GUI".

grorp reviewed (approved):
Tested together with #14211. Works well.

Thank you!

SmallJoker reviewed (approved):
Tested on Android 6.0. Works.
